src/installer/sources.py

```diff
@@ -122,6 +122,33 @@ def open(cls, path: "os.PathLike[str]") -> Iterator["WheelFile"]:
         with zipfile.ZipFile(path) as f:
             yield cls(f)
 
+    @property
+    def dist_info_dir(self) -> str:
+        """Name of the dist-info directory."""
+        if not hasattr(self, "_dist_info_dir"):
+            top_level_directories = {
+                path.split("/", 1)[0] for path in self._zipfile.namelist()
+            }
+            dist_infos = [
+                name for name in top_level_directories if name.endswith(".dist-info")
+            ]
+
+            assert (
+                len(dist_infos) == 1
+            ), "Wheel doesn't contain exactly one .dist-info directory"
+            dist_info_dir = dist_infos[0]
+
+            # NAME-VER.dist-info
+            di_dname = dist_info_dir.rsplit("-", 2)[0]
+            norm_di_dname = normalize_distribution_name(di_dname)
+            norm_file_dname = normalize_distribution_name(self.distribution)
+            assert (
+                norm_di_dname == norm_file_dname
+            ), "Wheel .dist-info directory doesn't match wheel filename"
```

Comment:
This isn't exactly the same check that pip does. It knows what distribution name it's expecting to install, so that's what it's looking for.

In the wheels I've seen that we hit this bug on, this assertion would be true.

Reply:
We also know the name, via `self.distribution`.

Reply:
Ah, right, in case of a mismatch. Nvm me.